processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -434,8 +434,49 @@ private SubclassMapping createSubclassMapping(SubclassMappingOptions subclassMap
         }
 
         private boolean isAbstractReturnTypeAllowed() {
-            return method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
-                && !method.getOptions().getSubclassMappings().isEmpty();
+            return !method.getOptions().getSubclassMappings().isEmpty()
+                && ( method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
+                    || isCorrectlySealed() );
+        }
+
+        private boolean isCorrectlySealed() {
+            try {
+                Type mappingSourceType = method.getMappingSourceType();
+                return isCorrectlySealed( mappingSourceType );
+            }
+            catch ( Exception e ) {
```

Comment:
What kind of an exception can happen here?

Reply:
true, reading the code there should be no chance of an exception happening here when not java 17.

Reply:
In that case let's remove the try catch. We should handle it properly if something throws an exception